### PR TITLE
Retarget to .NET 9

### LINKED ISSUE #10280: Retarget to .NET 9
Historically when we've hit releases for a new .NET version we've adopted that version. We should consider doing that for .NET 9 as well.

Items
- [ ] Retarget to Arcade  9 and build dor dotnet 9
- [ ] Update subscriptions to Microsoft.DotNet.Arcade.Sdk from `.NET ENG 8` To `".NET Eng - Latest" `
- [ ] https://github.com/dotnet/msbuild/issues/10557
- [ ] https://github.com/dotnet/msbuild/issues/9366
- [ ] Document the process of done work

### PR DESCRIPTION
**Pull request by f-alizada**
Fixes #10280

### Context
.NET 9 is in preview 6. 
Consuming the latest Arcade + retargeting the the msbuild core version to .net 9.

### Changes Made
- Upgrade the Target framework from .net8 to .net9
- `darc update-dependencies --channel ".NET Eng - Latest" --name "Microsoft.DotNet.Arcade.Sdk"`
- Version of xunit provided by arcade.sdk reverted, not to introduce the changes in one PR. 
- ApiCompat warnings supressed by `dotnet pack /p:GenerateCompatibilitySuppressionFile=true`
- Addressing affected test behavior by .NET 9 Preview 6 breaking change: https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/9.0/empty-env-variable. To delete the env var we need to pass the `null` value
- Addressed https://github.com/dotnet/msbuild/issues/9720
- Addressed https://github.com/dotnet/msbuild/issues/9366

### Testing
- Existing tests should pass
- Exp insertion: https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/569166

### CODE REVIEW

File: scripts/Deploy-MSBuild.ps1 [hunk 1]
```diff
@@ -76,7 +76,7 @@ else {
 if ($runtime -eq "Desktop") {
     $targetFramework = "net472"
 } else {
-    $targetFramework = "net8.0"
+    $targetFramework = "net9.0"
```
Comment (rainersigwald) [inline scripts/Deploy-MSBuild.ps1:79]: Ah, I forgot we haven't merged #10282! This shouldn't be necessary after that.

File: src/Tasks/ManifestUtil/ManifestReader.cs [hunk 2]
```diff
@@ -138,7 +142,10 @@ public static Manifest ReadManifest(string manifestType, string path, bool prese
             using (Stream s = File.Open(path, FileMode.Open, FileAccess.Read, FileShare.Read))
             {
                 byte[] buffer = new byte[2];
+#pragma warning disable CA2022 // Avoid inexact read with 'Stream.Read' The check of bytes happens later in the code. In case of invalid document the exception is expected later
```
Comment (rainersigwald) [inline src/Tasks/ManifestUtil/ManifestReader.cs:145]: Why did you choose to warning-disable these instead of doing it in editorconfig?

File: src/Tasks/ManifestUtil/PathUtil.cs [hunk 3]
```diff
@@ -162,7 +162,9 @@ public static bool IsPEFile(string path)
             byte[] buffer = new byte[2];
             using (Stream s = File.Open(path, FileMode.Open, FileAccess.Read, FileShare.Read))
             {
+#pragma warning disable CA2022 // Avoid inexact read with 'Stream.Read'
                 s.Read(buffer, 0, 2);
+#pragma warning restore CA2022 // Avoid inexact read with 'Stream.Read'
```
Comment (rainersigwald) [inline src/Tasks/ManifestUtil/PathUtil.cs:167]: Should we switch to `ReadExactly` and polyfill it on net472?

File: src/Utilities/README.md [hunk 4]
```diff
@@ -7,6 +7,6 @@ This package contains `Microsoft.Build.Utilities.Core.dll`, which defines helper
 * [`Logger`](https://docs.microsoft.com/dotnet/api/microsoft.build.utilities.logger), a base class for custom logging functionality.
 
 ### netstandard2.0 target
-The `netstandard2.0` target of this build is configured only to output reference assemblies; at runtime MSBuild will be `net8.0` or `net472`. Please use the `net8.0`-targeted assemblies for .NET 8+ scenarios.
+The `netstandard2.0` target of this build is configured only to output reference assemblies; at runtime MSBuild will be `net9.0` or `net472`. Please use the `net9.0`-targeted assemblies for .NET 8+ scenarios.
```
Comment (rainersigwald) [inline src/Utilities/README.md:10]:
```suggestion
The `netstandard2.0` target of this build is configured only to output reference assemblies; at runtime MSBuild will be `net9.0` or `net472`. Please use the `net9.0`-targeted assemblies for .NET 9+ scenarios.
```

File: src/Shared/FileMatcher.cs [hunk 5]
```diff
@@ -1673,7 +1673,7 @@ internal static bool IsMatch(string input, string pattern)
         /// <param name="pattern">Pattern against which string is matched.</param>
         internal static bool IsMatch(ReadOnlySpan<char> input, string pattern)
         {
-            if (input == null)
+            if (input == ReadOnlySpan<char>.Empty) // Comparing ReadOnlySpan with null in this scenario transforms the null to ReadOnlySpan Empty (https://github.com/dotnet/runtime/issues/84265)
```
Comment (rainersigwald) [inline src/Shared/FileMatcher.cs:1676]: I am glad to have this explanation and link now but I'm not sure it deserves a comment.

```suggestion
            if (input == ReadOnlySpan<char>.Empty)
```

File: src/Shared/FileMatcher.cs [hunk 6]
```diff
@@ -1673,7 +1673,7 @@ internal static bool IsMatch(string input, string pattern)
...
             {
                 throw new ArgumentNullException(nameof(input));
```
Comment (rainersigwald) [inline src/Shared/FileMatcher.cs:1678]: What's the right exception type to throw? Should this use `ThrowInternalError` instead?

File: .editorconfig [hunk 7]
```diff
@@ -413,5 +413,21 @@ dotnet_diagnostic.IDE0300.severity = suggestion
...
 # Temporarily disable SA1010 "Opening square brackets should not be preceded by a space" until https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3687 is fixed
 dotnet_diagnostic.SA1010.severity = none
+
+
+#TODO:fix all xUnit new warning 
+dotnet_diagnostic.xUnit1031.severity = suggestion
+dotnet_diagnostic.xUnit1048.severity = suggestion
+dotnet_diagnostic.xUnit2020.severity = suggestion
+dotnet_diagnostic.xUnit1012.severity = suggestion
+dotnet_diagnostic.xUnit2029.severity = suggestion
```
Comment (rainersigwald) [inline .editorconfig:433]: we disable one xunit analyzer in `eng/Common.globalconfig` already; move these there?

File: src/Build.UnitTests/BackEnd/BuildManager_Tests.cs [hunk 8]
```diff
@@ -1516,7 +1516,7 @@ public void CancelledBuildWithUnexecutedSubmission()
         /// <summary>
         /// A canceled build
         /// </summary>
-        [Fact(Timeout = 20_000)]
+        [Fact]
```
Comment (rainersigwald) [inline src/Build.UnitTests/BackEnd/BuildManager_Tests.cs:1519]: Could we migrate to the new xunit in a separate PR so everything's not quite so jumbled together?

File: src/Tasks/ManifestUtil/ManifestReader.cs (hunk 2 above)
Comment (f-alizada) [reply src/Tasks/ManifestUtil/ManifestReader.cs:145]: The only reason was: Turning off the warning globally will affect future contributions and will not report possible error/warnings

File: src/Tasks/ManifestUtil/PathUtil.cs (hunk 3 above)
Comment (f-alizada) [reply src/Tasks/ManifestUtil/PathUtil.cs:167]: For the scnarios where only 2 bytes are read, I decided to leave the s.Read and rely on the xml parser and proper exception handling. 
The plan is to use `ReadExactly` for cases when reading the files from 0 to filLength. and disable warning for net472.

File: src/Shared/FileMatcher.cs (hunk 6 above)
Comment (f-alizada) [reply src/Shared/FileMatcher.cs]: I reverted the change back to null detection, to resolve the failing tests first.  I'll investigate if we even need this check

File: scripts/Deploy-MSBuild.ps1 (hunk 1 above)
Comment (f-alizada) [reply scripts/Deploy-MSBuild.ps1:79]: the Deploy-MSBuild doesn't need to be updated? 
I have update that due to the fact that instructions of this script will be outdated if upgraded to the net9. 
cc: @YuliiaKovalova

File: .editorconfig (hunk 7 above)
Comment (f-alizada) [reply .editorconfig]: This is removed completely, since the version of xunit was set explicitly. 
And will be upgraded once this PR is merged

File: src/Shared/FileMatcher.cs (hunk 6 above)
Comment (f-alizada) [reply src/Shared/FileMatcher.cs]: Internal error added.

File: scripts/Deploy-MSBuild.ps1 (hunk 1 above)
Comment (f-alizada) [reply scripts/Deploy-MSBuild.ps1:79]: Closing, please feel free to reopen the comment if it is not the case

File: eng/Version.Details.xml [hunk 9]
```diff
@@ -98,7 +91,6 @@
     <Dependency Name="Microsoft.Net.Compilers.Toolset" Version="4.11.0-3.24378.3">
       <Uri>https://github.com/dotnet/roslyn</Uri>
       <Sha>5e3a11e2e7f952da93f9d35bd63a2fa181c0608b</Sha>
-      <SourceBuild RepoName="roslyn" ManagedOnly="true" />
```
Comment (MichaelSimons) [inline eng/Version.Details.xml:101]: Why was this removed?

File: global.json [hunk 10]
```diff
@@ -3,13 +3,13 @@
...
-    "dotnet": "8.0.201",
+    "dotnet": "9.0.100-preview.6.24328.19",
     "vs": {
       "version": "17.8.0"
     },
     "xcopy-msbuild": "17.8.5"
   },
   "msbuild-sdks": {
-    "Microsoft.DotNet.Arcade.Sdk": "8.0.0-beta.24376.1"
+    "Microsoft.DotNet.Arcade.Sdk": "9.0.0-beta.24379.1"
```
Comment (rainersigwald) [inline global.json:13]: Have we confirmed with dnceng that this is what they want for our LTS branch?

File: eng/Version.Details.xml (hunk 9 above)
Comment (f-alizada) [reply eng/Version.Details.xml:101]: After new arkade sdk was consumed, the source build was failing on this one, with errors: https://dev.azure.com/dnceng-public/public/_build/results?buildId=762782&view=logs&j=c316b76e-5b95-5e79-2740-62e92f13eebd&t=e33df8ad-5872-5105-0672-039423de0ed3 Only place <SourceBuild> elements on source-build intermediate nupkgs.

File: src/Build.UnitTests/Evaluation/ExpanderFunction_Tests.cs [hunk 11]
```diff
@@ -162,23 +162,26 @@ public void TryConvertToLongGivenDoubleWithLongMaxValueShouldNotThrow()
             _ = Should.NotThrow(() => Expander<IProperty, IItem>.Function<IProperty>.TryConvertToLong((double)long.MaxValue, out _));
         }
 
-        [Fact]
-        public void TryConvertToLongGivenDoubleWithLongMaxValue()
+        [WindowsFullFrameworkOnlyFact]
+        public void TryConvertToLongGivenDoubleWithLongMaxValueFramework()
```
Comment (rainersigwald) [inline src/Build.UnitTests/Evaluation/ExpanderFunction_Tests.cs:166]: I don't think I understand this test change. Can you elaborate on the reasoning?

File: src/Tasks/ResolveKeySource.cs [hunk 12]
```diff
@@ -127,7 +127,15 @@ private bool ResolveAssemblyKey()
                             fs = File.OpenRead(KeyFile);
                             int fileLength = (int)fs.Length;
                             var keyBytes = new byte[fileLength];
+#if NET7_0_OR_GREATER
+                            // fail fast in case the file is not read till the end
+                            fs.ReadExactly(keyBytes, 0, fileLength);
+#else
+#pragma warning disable CA2022 // Avoid inexact read with 'Stream.Read'
+                            // TODO: Read the count of read bytes and check if it matches the expected length, if not raise an exception
                             fs.Read(keyBytes, 0, fileLength);
```
Comment (rainersigwald) [inline src/Tasks/ResolveKeySource.cs:136]: I'd still prefer a polyfill here so the source can just unconditionally say `ReadExactly`

File: src/Tasks/CompatibilitySuppressions.xml [hunk 13]
```diff
@@ -1,10 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- https://learn.microsoft.com/en-us/dotnet/fundamentals/package-validation/diagnostic-ids -->
 <Suppressions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
-<!-- For ease of logging the "not supported on Core" message, these tasks are a
-         TaskRequiresFramework on netstandard/netcore. Since the type is sealed there,
-         that shouldn't cause any implementation problems since no one can derive
-         from it and try to call TaskExtension.Log. -->
```
Comment (rainersigwald) [inline src/Tasks/CompatibilitySuppressions.xml:7]: Please preserve the comments where the suppressions aren't removed.

File: src/Tasks/CompatibilitySuppressions.xml [hunk 14]
```diff
@@ -71,12 +67,55 @@
     <Left>ref/netstandard2.0/Microsoft.Build.Tasks.Core.dll</Left>
     <Right>ref/net472/Microsoft.Build.Tasks.Core.dll</Right>
   </Suppression>
-
-  <!-- PKV004 for netstandard2.0-supporting TFs that we do not have runtime assemblies for.
-       This is intentional, because you can only use MSBuild in the context of a .NET SDK
-       (on net7.0, as of MSBuild 17.4) or in the context of Visual Studio (net472), but we
-       have previously shipped netstandard2.0 packages, and if you want to support both
-       runtime contexts it still makes sense to target that. -->
+  <Suppression>
```
Comment (rainersigwald) [inline src/Tasks/CompatibilitySuppressions.xml:70]: Please _add_ a comment for these new suppressions.

File: Directory.Build.props [hunk 15]
```diff
@@ -87,4 +87,13 @@
...
 
+  <!--  Freeze the version for the current upgrade of the sdk. 
+        TODO:Delete the lines below once the Retargeting merged. 
+  -->  
+  <PropertyGroup>
+      <XUnitVersion>2.4.2</XUnitVersion>
+      <XUnitAnalyzersVersion>1.1.0</XUnitAnalyzersVersion>
+      <XUnitRunnerConsoleVersion>$(XUnitVersion)</XUnitRunnerConsoleVersion>
+      <XUnitRunnerVisualStudioVersion>2.4.3</XUnitRunnerVisualStudioVersion>
+  </PropertyGroup>
```
Comment (rainersigwald) [inline Directory.Build.props:98]: Can these go to `Versions.props`? that's where I'd normally look.

File: eng/Version.Details.xml (hunk 9 above)
Comment (MichaelSimons) [reply eng/Version.Details.xml:101]: The correct fix is to add the corresponding soucebuild intermediate dependency (ideally right after this dependency)

```
    <Dependency Name="Microsoft.SourceBuild.Intermediate.roslyn" Version="4.11.0-3.24378.3">
      <Uri>https://github.com/dotnet/roslyn</Uri>
      <Sha>5e3a11e2e7f952da93f9d35bd63a2fa181c0608b</Sha>
      <SourceBuild RepoName="roslyn" ManagedOnly="true" />
    </Dependency>
```

Then delete the change to https://github.com/dotnet/msbuild/pull/10451/files#diff-f88ce4026366ce74fb1a981785c542757617abf4393a33ca34465a5e9a0dd343

File: src/Build/CompatibilitySuppressions.xml [hunk 16]
```diff
@@ -128,17 +72,7 @@
...
-    <IsBaselineSuppression>true</IsBaselineSuppression>
-  </Suppression>
-  <Suppression>
-    <DiagnosticId>CP0011</DiagnosticId>
-    <Target>F:Microsoft.Build.Experimental.BuildCheck.BuildAnalyzerResultSeverity.Warning</Target>
-    <Left>ref/net8.0/Microsoft.Build.dll</Left>
-    <Right>ref/net8.0/Microsoft.Build.dll</Right>
-    <IsBaselineSuppression>true</IsBaselineSuppression>
+    <DiagnosticId>PKV006</DiagnosticId>
+    <Target>net8.0</Target>
```
Comment (JanKrivanek) [inline src/Build/CompatibilitySuppressions.xml:76]: Is this right? How can net8.0 it occur?

File: src/Tasks/ResolveKeySource.cs (hunk 12 above)
Comment (JanKrivanek) [reply src/Tasks/ResolveKeySource.cs:136]: Let's track that than. It feels that it'd be outside of scope of retargetting change.

File: global.json (hunk 10 above)
Comment (f-alizada) [reply global.json:13]: https://github.com/dotnet/dnceng/issues/3713

**Review by JanKrivanek — COMMENTED** [review]
Excluding the changes in eng folder (I haven't reviewed those) and the commented stuff it looks OK.
The VS insertion looks OK as well (both tests failures are expected and known, and this change even brings some nice perf improvements).

The only thing missing for me for signoff (appart from commented stuff) is a list of manual changes to eng folder that has been done beyond arcade consuming (if any beyond already described and https://github.com/dotnet/msbuild/pull/10451#discussion_r1702081041)

File: src/Build/CompatibilitySuppressions.xml (hunk 16 above)
Comment (f-alizada) [reply src/Build/CompatibilitySuppressions.xml:76]: The Diagnostic id is PKV006. 
> The target framework is dropped in the latest version.

https://learn.microsoft.com/en-us/dotnet/fundamentals/apicompat/diagnostic-ids

**Comment by rainersigwald** [comment]
> Excluding the changes in eng folder

You can't skip `eng/`, only `eng/common`! We have scary stuff in there :)

File: src/Build.UnitTests/Evaluation/ExpanderFunction_Tests.cs (hunk 11 above)
Comment (f-alizada) [reply src/Build.UnitTests/Evaluation/ExpanderFunction_Tests.cs:166]: The test previously introduced here https://github.com/dotnet/msbuild/commit/0f6ed65aea401b96a639c8b84db7d0688a48591d
Testing different behaviour on different archtitecture. 
However after moving to .net 9 the precision is not lost hence we are receveing the expected outcome of what was written: 

```
result.ShouldBeTrue();
actual.ShouldBe(longMaxValue);
```
This is true only for .NET 9 target, and the Framework left as it is.

**Comment by f-alizada** [comment]
Closing this PR with respect to the new PR with more clear history and less changes: https://github.com/dotnet/msbuild/pull/10484
A lot of learnings from here :)
